include/sys/zstd/zstd.h

```diff
@@ -94,6 +97,123 @@ int zfs_zstd_decompress(void *s_start, void *d_start, size_t s_len,
     size_t d_len, int n);
 void zfs_zstd_cache_reap_now(void);
 
+/*
+ * So, the reason we have all these complicated set/get functions is that
+ * originally, in the zstd "header" we wrote out to disk, we used a 32-bit
+ * bitfield to store the "level" (8 bits) and "version" (24 bits).
+ *
+ * Unfortunately, bitfields make few promises about how they're arranged in
+ * memory...
+ *
+ * As a consequence, we wound up with little endian platforms with a layout
+ * like this in memory:
+ *
+ *      32      24      16      8       0
+ *      +-------+-------+-------+-------+
+ *      | level |   0   |    version    |
+ *      +-------+-------+-------+-------+
+ *
+ * ...and then, after being run through BE_32(), serializing this out to
+ * disk:
+ *
+ *      32      24      16      8       0
+ *      +-------+-------+-------+-------+
+ *      |    version    |   0   | level |
```

Comment:
I think that this logic depends on "version" having some bits set in both its low byte and its high byte.  It looks like that's the case for the current ZSTD_VERSION_NUMBER, but could we VERIFY() that's the case before we write it out?  Otherwise we might accidentally bump to version `1.05.12`, which I think would have a zero low byte, and then this logic breaks.

Reply:
You're correct, it does, though I don't think 1.5.12 would do it? If I'm understanding ZSTD_VERSION_NUMBER right, 1.5.12 turns into 1*10000 + 5*100 * 12 = 10512/0x2910 - 1.4.96 would, though, or 1.7.52, or ...

So (unless I'm overlooking something) there's two cases that can happen without the version number rising pretty dramatically far - the low byte is 0x00, or the high byte is 0x00. One effectively can't happen without the version rolling back in time to < 0.02.56 (or forward past 6.55.35, at which point I'd hope we could either drop the compat code or put it behind a non-default tunable), and the other winds up with a layout like 00 v0 00 LV or LV 00 v0 00; in either of the latter two cases, I believe the code will do the right thing no matter which order it iterates (though it should be consistent in order across platforms with this version anyway).